### PR TITLE
feat: validate atm profile when handling backend update

### PR DESCRIPTION
**Pull request by zhiying-lin**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

validate whether the atm profile and populate the backend status

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:

- [X] run `make reviewable` for basic local test
- [ ] Read and followed fleet-networking's [Code of conduct](https://github.com/Azure/fleet-networking/blob/main/CODE_OF_CONDUCT.md).
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests

### CODE REVIEW

File: pkg/controllers/hub/trafficmanagerbackend/controller.go [hunk 1]
```diff
@@ -246,6 +256,35 @@ func (r *Reconciler) validateTrafficManagerProfile(ctx context.Context, backend
...
+		if azureerrors.IsNotFound(getErr) {
+			// We've already checked the TrafficManagerProfile condition before getting Azure resource.
+			// It may happen when
+			// 1. customers delete the azure profile manually
+			// 2. the TrafficManagerProfile info is stale.
+			// For the case 1, retry won't help to recover the Azure Traffic Manager profile resource.
+			// For the case 2, the controller will be re-triggered when the TrafficManagerProfile is updated.
+			klog.ErrorS(getErr, "NotFound Azure Traffic Manager profile", "trafficManagerBackend", backendKObj, "trafficManagerProfile", profileKObj, "atmProfileName", atmProfileName)
+			// none of the endpoints are accepted by the TrafficManager
+			setFalseCondition(backend, nil, fmt.Sprintf("Azure Traffic Manager profile %q is not found", atmProfileName))
```
Comment (jwtty) [inline pkg/controllers/hub/trafficmanagerbackend/controller.go:275]:
```suggestion
			setFalseCondition(backend, nil, fmt.Sprintf("Azure Traffic Manager profile %s/%s is not found", r.ResourceGroupName, atmProfileName))
```

File: test/common/trafficmanager/validator/backend.go [hunk 2]
```diff
@@ -67,3 +67,14 @@ func IsTrafficManagerBackendDeleted(ctx context.Context, k8sClient client.Client
 		return nil
 	}, timeout, interval).Should(gomega.Succeed(), "Failed to remove trafficManagerBackend %s ", name)
 }
+
+// ConsistentlyTrafficManagerConsistentlyExist validates whether the backend consistently exists.
+func ConsistentlyTrafficManagerConsistentlyExist(ctx context.Context, k8sClient client.Client, name types.NamespacedName) {
```
Comment (jwtty) [inline test/common/trafficmanager/validator/backend.go:72]: `ConsistentlyTrafficManagerConsistentlyExist` seems weird? Are we following any naming convention?

File: pkg/controllers/hub/trafficmanagerbackend/suite_test.go [hunk 3]
```diff
@@ -54,7 +55,9 @@ func TestAPIs(t *testing.T) {
 }
 
 var _ = BeforeSuite(func() {
-	klog.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
+	logger := zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true))
+	klog.SetLogger(logger)
+	log.SetLogger(logger)
```
Comment (jwtty) [inline pkg/controllers/hub/trafficmanagerbackend/suite_test.go:60]: Why do we need this change?

File: pkg/controllers/hub/trafficmanagerprofile/suite_test.go [hunk 4]
```diff
@@ -53,7 +54,9 @@ func TestAPIs(t *testing.T) {
 }
 
 var _ = BeforeSuite(func() {
-	klog.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
+	logger := zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true))
+	klog.SetLogger(logger)
+	log.SetLogger(logger)
```
Comment (jwtty) [inline pkg/controllers/hub/trafficmanagerprofile/suite_test.go:59]: +1 Why do we need this change?

File: pkg/controllers/hub/trafficmanagerbackend/controller_integration_test.go [hunk 5]
```diff
@@ -108,8 +129,31 @@ var _ = Describe("Test TrafficManagerBackend Controller", func() {
 			Expect(k8sClient.Create(ctx, backend)).Should(Succeed())
 		})
 
+		It("Updating TrafficManagerProfile status to accepted true and it should trigger controller", func() {
+			By("By updating TrafficManagerProfile status")
```
Comment (jwtty) [inline pkg/controllers/hub/trafficmanagerbackend/controller_integration_test.go:133]: Without adding this condition status, what's gonna happen? The trafficManagerProfile will fail right?

File: pkg/controllers/hub/trafficmanagerbackend/controller_integration_test.go [hunk 6]
```diff
@@ -108,8 +129,31 @@ var _ = Describe("Test TrafficManagerBackend Controller", func() {
 			Expect(k8sClient.Create(ctx, backend)).Should(Succeed())
 		})
 
+		It("Updating TrafficManagerProfile status to accepted true and it should trigger controller", func() {
```
Comment (jwtty) [inline pkg/controllers/hub/trafficmanagerbackend/controller_integration_test.go:132]:
```suggestion
		It("Updating TrafficManagerProfile status to programmed true and it should trigger controller", func() {
```

File: pkg/controllers/hub/trafficmanagerbackend/controller_integration_test.go [hunk 7]
```diff
@@ -131,6 +175,85 @@ var _ = Describe("Test TrafficManagerBackend Controller", func() {
...
+		backendNamespacedName := types.NamespacedName{Namespace: testNamespace, Name: backendName}
+		var backend *fleetnetv1alpha1.TrafficManagerBackend
+
+		It("Creating a new TrafficManagerProfile", func() {
+			By("By creating a new TrafficManagerProfile")
+			profile = trafficManagerProfileForTest(profileName)
+			Expect(k8sClient.Create(ctx, profile)).Should(Succeed())
+		})
+
+		It("Updating TrafficManagerProfile status to accepted true", func() {
```
Comment (jwtty) [inline pkg/controllers/hub/trafficmanagerbackend/controller_integration_test.go:192]:
```suggestion
		It("Updating TrafficManagerProfile status to programmed true", func() {
```

File: pkg/controllers/hub/trafficmanagerbackend/controller_integration_test.go [hunk 8]
```diff
@@ -145,6 +268,18 @@ var _ = Describe("Test TrafficManagerBackend Controller", func() {
 			Expect(k8sClient.Create(ctx, profile)).Should(Succeed())
 		})
 
+		It("Updating TrafficManagerProfile status to accepted true", func() {
```
Comment (jwtty) [inline pkg/controllers/hub/trafficmanagerbackend/controller_integration_test.go:271]:
```suggestion
		It("Updating TrafficManagerProfile status to programmed true", func() {
```

File: pkg/controllers/hub/trafficmanagerbackend/controller_integration_test.go [hunk 9]
```diff
@@ -187,6 +322,18 @@ var _ = Describe("Test TrafficManagerBackend Controller", func() {
...
```
Comment (jwtty) [inline pkg/controllers/hub/trafficmanagerbackend/controller_integration_test.go:325]:
```suggestion
		It("Updating TrafficManagerProfile status to programmed true", func() {
```

File: pkg/controllers/hub/trafficmanagerbackend/suite_test.go (hunk 3 above)
Comment (zhiying-lin) [reply pkg/controllers/hub/trafficmanagerbackend/suite_test.go:60]: the log complains that the setLogger is never called. so i made this change

File: pkg/controllers/hub/trafficmanagerbackend/controller_integration_test.go (hunk 5 above)
Comment (zhiying-lin) [reply pkg/controllers/hub/trafficmanagerbackend/controller_integration_test.go]: added the validation, the condition will be set to unknown

**Review by jwtty — APPROVED** [review]

**Pull request merged**